### ISSUE TITLE
Episode 3: monitoring the training process 

### ISSUE THREAD

**Issue by dafnevk** [carpentries-incubator/deep-learning-intro#44]
## Learning objectives: 
- Explain the importance of splitting the training data 
- Use the data splits to plot the training process 
- Measure the performance of your deep neural  network 
- Interpret the training plots to recognize overfitting 
- Implement basic strategies to prevent overfitting 
- Understand the effect of regularization techniques

**Comment by florian-huber** [comment]
### Additional things to coordinate:
- Ideally use data that was also used in Episode 2 (#43)
- Make consistent use of Keras API across episodes.

**Comment by florian-huber** [comment]
Regarding datasets:
- I did run some trials with the Iris data set, but I probably would prefer to chose another one. The Iris data is too easy to fit (most architectures with 2 or more layers will simply reach nearly 100% accuracy after a few epochs) and it doesn't contain enough samples (150 samples, making it a bit hard to create proper test and validation sets). 
- As a next step I will look into the [One-hundred plant species leaves data set](https://archive.ics.uci.edu/ml/datasets/One-hundred+plant+species+leaves+data+set)hundred+plant+species+leaves+data+set), the [Mushroom data set](https://archive.ics.uci.edu/ml/datasets/mushroom) which both would be classification problems as well.
- Another alternative would be to work on a regression problem, e.g using the bike rental dataset, or the Portugal forest fire dataset.

**Comment by psteinb** [comment]
I like the idea behind the mushroom data set (even though I haven't looked at it) as it can set the scene for ethical considerations/discussions.

**Comment by florian-huber** [comment]
Short update:
What I mostly did so far is to explore various datasets, looking for a simple+comprehensible but also interesting enough dataset to show the effect of overfitting.
- **Mushroom data set** (predict toxic/not-toxic): 
See notebook [here](https://github.com/carpentries-incubator/deep-learning-intro/blob/epidode3-prototyping/_episodes/episode3_dataset_exploration/DNN_mushroom_classifier_subset.ipynb). The data is again too simple for deep learning I think. Even when I only use a tiny fraction of the given features AND a small fraction of all samples, there's hardly any overfitting. Maybe I could use even less samples...?
- **Forest fire dataset** (Predict forest fire and its size based on date and weather data). Here I do see overfitting more easily. But: The dataset is not so straightforward. It is a very difficult regression problem, because on many days there simply is no fire at all. So you would have to combine a classifier (is there a fire - yes/no) with a regression (if yes, what's the area). Could probably be made feasible by taking out the events without fire and simplifying the rest.

**Comment by florian-huber** [comment]
I have now created a new dataset which combined daily weather data from 18 location in Europe between 2000 and 2010. That can be used to do regression (predict, say, sunshine hours for one of the locations based on the 17 others) or to do classification (can a model predict the month based on the weather data of a single day).

**Comment by dafnevk** [comment]
That sounds like a useful dataset indeed. Have you hosted it somewhere, so that we can add the download link to the installation instructions?

**Comment by florian-huber** [comment]
I still need to add all relevant metadata and then I was planning to upload it to zenodo.

**Comment by florian-huber** [comment]
Hi. So here an update. Recently, @dafnevk, @bpmweel and me gave small deep learning lesson in which we used episode 2 (#67) and a first part of episode 3.

Some of the conclusion were:
- episode 1 & 2 gave a good introduction to the topic and a first introduction to keras
- in episode 3 people will have to front the task of building their first neural network from scratch (largely following the example from episode 2, but still). This needs time and room for questions/discussions.
- the current material contains to much extras which are not realistic to teach in one episode, so it needs a bit more cleaning up (or moving some things to an optional episode, e.g. topics such as model uncertainty and handling unbalanced data).
- We could even decide to split this episode in two:
(3a) same learning goals as mentioned by Dafne on top, but leaving out regularization techniques (overfitting will be shown and simply be avoided using "early stopping")
(3b) ways to prevent or handle overfitting and improving the model (regularization techniques such as dropout, batchnorm, ...) ?
